File: Boxuan/16_01_18/Find_Peak_Element.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        if (nums.size() == 0){
            return -1;
        }
        int left = 0;
        int right = nums.size() - 1;
        while (left + 1 < right){
            int mid = left + (right - left) / 2;
            if (nums[mid] > nums[mid + 1] and nums[mid] > nums[mid - 1]){
                return mid;
            }
            else if (nums[mid] > nums[mid + 1]){
                right = mid - 1;
            }
            else{
                left = mid + 1;
            }
        }
        int ans = nums[right] > nums[left]?right:left;
        return ans;
    }
};

File: Boxuan/16_01_18/Find_Peak_Element.py
class Solution(object):
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return -1
            
        left, right = 0, len(nums) - 1
        while left + 1 < right:
            mid = left + (right - left) / 2
            if nums[mid] > nums[mid - 1] and nums[mid] > nums[mid + 1]:
                return mid
            if nums[mid] < nums[mid - 1]:
                right = mid - 1 
            elif nums[mid] < nums[mid + 1]:
                left = mid + 1
        ans = left if nums[left] > nums[right] else right
        return ans